papis/downloaders/usenix.py

```diff
@@ -80,26 +89,75 @@ def get_bibtex_url(self) -> Optional[str]:
         o = urlparse(self.uri)
         import bs4
 
+        # make sure self._raw_data is available
+        self._ensure_raw_data()
         if not self._raw_data:
-            self._raw_data = self.session.get(self.uri).content.decode("utf-8")
-            if not self._raw_data:
-                self.logger.warning("Failed to fetch data from '%s'.", self.uri)
-                return None
+            return None
 
         soup = bs4.BeautifulSoup(self._raw_data, "html.parser")
+        re_matcher = re.compile(r"/biblio/export/bibtex/([0-9]+)$")
 
         a = list(
             filter(
-                lambda t: re.match(r"/biblio/export/bibtex/([0-9]+)$",
-                                   t.get("href", "")),
+                lambda t: re_matcher.match(t.get("href", "")),
                 soup.find_all("a"),
-            ))
+            )
+        )
 
         if not a:
-            self.logger.warning("No BibTeX URL found in this usenix page: '%s'.",
-                                self.uri)
+            self.logger.warning(
+                "No BibTeX URL found in this usenix page: '%s'.", self.uri
+            )
             return None
 
         bib_path = a[0].get("href", "")
         bib_url = o._replace(path=bib_path)
         return bib_url.geturl()
+
+    def download_bibtex(self) -> None:
+        """Download and store that BibTeX data from :meth:`get_bibtex_url`.
+        If that doesn't work, e.g., because cloudflare doesn't like papis for
+        some reason, try to find the inline bibtex content and use that instead.
+
+        Use :meth:`get_bibtex_data` to access the metadata from the BibTeX URL.
+        """
+        url = self.get_bibtex_url()
+        if not url:
+            return
+        self.logger.info("Downloading BibTeX from '%s'.", url)
+
+        response = self.session.get(url, cookies=self.cookies)
+        self.bibtex_data = response.content.decode().strip()
+        if self.bibtex_data.startswith("<!DOCTYPE html>"):
+            self.logger.debug("downloaded bibtex data: %s", self.bibtex_data)
+            self.bibtex_data = None
+
+        if self.bibtex_data:
+            return
+
+        # fallback to trying to fetch the bibtex from the _raw_data itself
+        # make sure self._raw_data is available
+        self._ensure_raw_data()
+        if not self._raw_data:
+            return None
+        # setup html parser
+        import bs4
+
+        soup = bs4.BeautifulSoup(self._raw_data, "html.parser")
```

Comment:
Does `self._get_soup()` not work here? I think that defaults to an `lxml` parser.

Reply:
Hm, agreed, but feels like this could be reworked to take advantage of it (`self._get_soup` is also cached), but it's not a pressing issue.